aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,137 @@
+{
+    // get the current out_offs location
+    uint8_t out_offs = buffer->out_offs;
+
+    // while char_offset is greater than size of the current entry
+    // reduce offset by size of current entry
+    // advance the out_offs location
+    while(char_offset >= buffer->entry[out_offs].size)
+    {
+        char_offset -= buffer->entry[out_offs].size;
+
+        // advance the out_offs location
+        if(out_offs == (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1))
+        {
+            out_offs = 0;
+        }
+        else
+        {
+            out_offs++;
+        }
+        
+        // check if offset exceeds last byte
+        if(out_offs == buffer->in_offs)
+        {
+            if(char_offset <= 0)
+            {
+                return NULL;
+            }
+        }
+
+        // if entry doesn't exist
+        // return NULL for no data found
+        if(buffer->entry[out_offs].buffptr == NULL)
+        {
+            return NULL;
+        }
+    }
+    
+    // if entry doesn't exist
+    // return NULL for no data found
+    if(buffer->entry[out_offs].buffptr == NULL)
+    {
+        return NULL;
+    }
+
+    // store byte of the returned aesd_buffer_entry->buffptr member
+    // corresponding to char_offset.
+    *entry_offset_byte_rtn = char_offset;
+
+    // return buffer entry
+    return &(buffer->entry[out_offs]);
+}
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    // when buffer is not full
+    if(buffer->full == false)
+    {
+        // store buffer entry
+        buffer->entry[buffer->in_offs].buffptr = add_entry->buffptr;
+        buffer->entry[buffer->in_offs].size = add_entry->size;
+
+        // check if in_offs location has reached end of buffer
+        if(buffer->in_offs == (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1))
+        {
+            // go back to start of buffer
+            buffer->in_offs = 0;
+        }
+        else
+        {
+            // increment in_offs location
+            buffer->in_offs++;
+        }
+
+        // check and set if buffer is full
+        if(buffer->in_offs == buffer->out_offs)
+        {
+            buffer->full = true;
+        }
+    }
+    // when buffer is full
+    else if(buffer->full == true)
+    {
+        // store buffer entry
+        buffer->entry[buffer->in_offs].buffptr = add_entry->buffptr;
+        buffer->entry[buffer->in_offs].size = add_entry->size;
+
+        // increment in_offs and out_offs location
+        buffer->in_offs++;
+        buffer->out_offs++;
```

Comment:
@ameyflash is this safe to do?  What would happen if `in_offs == out_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1`?